src/mono/mono/mini/interp/mintops.h

```diff
@@ -235,9 +235,10 @@ typedef enum {
 #define MINT_IS_LDIND_OFFSET(op) ((op) >= MINT_LDIND_OFFSET_I1 && (op) <= MINT_LDIND_OFFSET_I8)
 #define MINT_IS_SIMD_CREATE(op) ((op) >= MINT_SIMD_V128_I1_CREATE && (op) <= MINT_SIMD_V128_I8_CREATE)
 #define MINT_IS_RETURN(op) (((op) >= MINT_RET && (op) <= MINT_RET_U2) || (op) == MINT_RET_I4_IMM || (op) == MINT_RET_I8_IMM)
+#define MINT_IS_BOX(op) ((op) == MINT_BOX || (op) == MINT_BOX_VT || (op) == MINT_BOX_PTR || (op) == MINT_BOX_NULLABLE_PTR)
 
 // TODO Add more
-#define MINT_NO_SIDE_EFFECTS(op) (MINT_IS_MOV (op) || MINT_IS_LDC_I4 (op) || MINT_IS_LDC_I8 (op) || op == MINT_LDC_R4 || op == MINT_LDC_R8 || op == MINT_LDPTR || op == MINT_BOX || op == MINT_INITLOCAL)
+#define MINT_NO_SIDE_EFFECTS(op) (MINT_IS_MOV (op) || MINT_IS_LDC_I4 (op) || MINT_IS_LDC_I8 (op) || op == MINT_LDC_R4 || op == MINT_LDC_R8 || op == MINT_LDPTR || MINT_IS_BOX(op) || op == MINT_INITLOCAL)
```

Comment:
It is a bit dubious indeed. According to our tests, a pattern like `box + ldnull + ceq` will always be optimized out, even if the box would normally throw. However `box + pop` should preserve the throw. Rather than hardcode IL patterns I figured it would be better to always attempt to optimize out box of byreflike, also reusing optimization for non-byreflike boxes. While it is not 100% clear to me what the rules are, I'm merging this as is since it should allow usage of this feature and the relevant tests are passing.